python/cog/director/monitor.py

```diff
@@ -0,0 +1,93 @@
+import threading
+import time
+from queue import Empty, Full, Queue
+from typing import Optional
+
+import structlog
+from opentelemetry import trace
+
+from .. import schema
+
+log = structlog.get_logger(__name__)
+
+
+class Monitor:
+    def __init__(self, utilization_interval=15):
+        self._thread = None
+        self._should_exit = threading.Event()
+        self._tracer = trace.get_tracer("cog-director")
+        self._prediction_events = Queue()
+        self.utilization_interval = utilization_interval
+
+    def start(self) -> None:
+        self._thread = threading.Thread(target=self._run)
+        self._thread.start()
+
+    def stop(self) -> None:
+        self._should_exit.set()
+
+    def join(self) -> None:
+        if self._thread is not None:
+            self._thread.join()
+
+    def set_current_prediction(self, prediction: schema.PredictionResponse):
+        try:
+            self._prediction_events.put_nowait(prediction)
+        except Full:
+            log.info("prediction event queue is full, dropping event")
+
+    def _run(self) -> None:
+        last_span_at = time.perf_counter() - self.utilization_interval
+        active_sec = 0.0
+        current_prediction: Optional[schema.PredictionResponse] = None
+        current_prediction_started_at: Optional[float] = None
+
+        while not self._should_exit.is_set():
+            try:
+                current_prediction = self._prediction_events.get_nowait()
+                # Set to None or schema.PredictionResponse
+                # Will raise Empty when there is nothing in the queue in the timeout
+
+                if current_prediction_started_at:  # count utilization in window
+                    active_sec += time.perf_counter() - current_prediction_started_at
+
+                if current_prediction:  # new prediction
+                    current_prediction_started_at = time.perf_counter()
+                else:  # switched to idling
+                    current_prediction_started_at = None
+            except Empty:
+                pass
+
+            if time.perf_counter() >= last_span_at + self.utilization_interval:
+                with self._tracer.start_as_current_span(
+                    name="cog.director.utilization"
+                ) as span:
+                    if current_prediction_started_at:  # event didn't finish in window
+                        active_sec += (
+                            time.perf_counter() - current_prediction_started_at
+                        )
+
+                    utilization_for_window = min(
+                        active_sec / self.utilization_interval, 1.0
+                    )
```

Comment:
The line length configuration for black really makes this way harder to read than it should be... I think we should move to 120 for the repo